src/JuMP.jl

```diff
@@ -938,24 +938,65 @@ end
 
 """
     get_optimizer_attribute(
-        model::Model, attr::MOI.AbstractOptimizerAttribute
+        model::Model,
+        attr::MOI.AnyAttribute,
+        args...
     )
 
 Return the value of the solver-specific attribute `attr` in `model`.
 
-## Example
+!!! warning
+    If you are querying a solver-specific attribute such as
+    `Gurobi.ModelAttribute`, and you are not using [`direct_model`](@ref), and
+    you have not called [`optimize!`](@ref), you must first call
+    `MOI.Utilities.attach_optimizer(model)` in order to first copy the problem
+    into the solver.
+
+See also: [`set_optimizer_attribute`](@ref), [`set_optimizer_attributes`](@ref).
+
+## Examples
 
+Query model or optimizer attributes:
 ```julia
 get_optimizer_attribute(model, MOI.Silent())
 ```
 
-See also: [`set_optimizer_attribute`](@ref), [`set_optimizer_attributes`](@ref).
+Query variable or constraint attributes:
+```julia
+get_optimizer_attribute(model, MOI.VariablePrimalStart(), x)
+```
+
+Query a solver-specific attribute:
+```julia
+MOI.Utilities.attach_optimizer(model)
+get_optimizer_attribute(model, Gurobi.ModelAttribute("IsMIP"))
+```
 """
+function get_optimizer_attribute(model::Model, attr::MOI.AnyAttribute, args...)
+    # If we're in direct mode, no need for `AttributeFromOptimizer`.
+    if mode(model) == DIRECT
+        return MOI.get(model, attr, args...)
+    end
+    # ... else we have a caching optimizer.
+    state = MOI.Utilities.state(backend(model))
+    if state == MOI.Utilities.NO_OPTIMIZER ||
+       state == MOI.Utilities.EMPTY_OPTIMIZER
+        # Query from the model cache.
+        return MOI.get(model, attr, args...)
+    else
+        @assert state == MOI.Utilities.ATTACHED_OPTIMIZER
+        new_attr = MOI.Utilities.AttributeFromOptimizer(attr)
```

Comment:
`AttributeFromOptimizer` will allow you to go through one CachingOptimizer layer but if there is a second one then you will query from its cache.

Reply:
It seems we need to do a recursion through the MOI layers at the JuMP level and not use `AttributeFromOptimizer`. Why not use `MOI.get(unsafe_backend(model), attr, ...)` ? By the way, there is the same discussion about having to call attach first there.

Reply:
How about `unsafe_get_optimizer_attribute` if you want to query solver-specific attributes that you haven't set yet?

Reply:
Keeping also the method of this PR?